### DISCUSSION TITLE
PictureBox Image property API work-around

### DISCUSSION THREAD

**Discussion by LastJasonNext** [twinbasic/twinbasic#1535] [Q&A]
Form and PictureBox Image property is yet to be implemented. Does anyone have an API workaround?

**Comment by fafalone** [comment]
What specific difference are you hoping to replicate that's not possible with the `.Picture` properties, which are implemented?

**Reply by LastJasonNext** [reply]
Thanks for the question, Jon. I have used .Print and .Line to place content within the PictureBox. In order to create a .Picture of that image, for the same or other .Picture equipped control, the .Image property creates a bitmap. So, control.Picture = PictureBox.Image literally captures what's been Printed or Lined in the PictureBox.

-Jason

**Comment by fafalone** [comment]
You could use the `PrintWindow` API to copy the image to an hDC you provide, then whatever from there.

**Reply by wqweto** [reply]
Note that you can retrieve `PictureBox.Image` only when `AutoRedraw` is set i.e. `Image` property allows access to the `hBmp` used for double-buffering of WM_PAINT/WM_PRINTCLIENT output.

Without `AutoRedraw` all Circle/Lines are not persisted and get "erased" on WM_PAINT which in turn uses control's `BackColor` and `Picture` to redraw in case when no double-buffer bitmap is available.

Such interaction does not allow `PrintWindow` (nor WM_PRINTCLIENT) to "snapshot" current output on display as all Circle/Line primitives would be "lost".

**Comment by LastJasonNext** [comment]
Thank you. My VBP sets AutoRedraw in the project's use of
PictureBox.Image. Jon mentioned using PrintWindow, which is the API
workaround I was looking for, and your response elucidated it's usage
beautifully. Meanwhile, while a work around would bring the VBP closer
to being able to work within TwinBASIC's IDE, patience is the virtue of
choice. William and Company are checking off the unimplemented list with
a measured alacrity. What a project. I'm in a bit of awe here. 

On 2023-04-15 04:17, Vladimir Vissoultchev wrote:

> Note that you can retrieve PictureBox.Image only when AutoRedraw is set i.e. Image property allows access to the hBmp used for double-buffering of WM_PAINT/WM_PRINTCLIENT output. 
> 
> Without AutoRedraw all Circle/Lines are not persisted and get "erased" on WM_PAINT which in turn uses control's BackColor and Picture to redraw in case when no double-buffer bitmap is available. 
> 
> Such interaction does not allow PrintWindow (nor WM_PRINTCLIENT) to "snapshot" current output on display as all Circle/Line primitives would be "lost". 
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe [2].
> You are receiving this because you authored the thread.Message ID: ***@***.***>
 

Links:
------
[1]
https://github.com/twinbasic/twinbasic/discussions/1535#discussioncomment-5623049
[2]

**Comment by LastJasonNext** [comment]
Thanks, Jon, that was the workaround I was searching for. Meanwhile,
patience is the real solution. The unimplemented list is shrinking. What
a project! 

On 2023-04-14 14:18, Jon wrote:

> You could use the PrintWindow API to copy the image to an hDC you provide, then whatever from there. 
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe [2].
> You are receiving this because you authored the thread.Message ID: ***@***.***>
 

Links:
------
[1]
https://github.com/twinbasic/twinbasic/discussions/1535#discussioncomment-5620323
[2]

**Comment by DMacPhersonHCI** [comment]
I'd like to add my interest in getting the Image object in the PictureBox. We use this for coordinate line plot creation, and the drawing piece of it works great. But we need the Image property to export with SavePicture. (The PrintWindow is not sufficient as many of these are constructed in background hidden forms.) I've tried to find a workaround, but the persistent-image property is the one missing piece. Thanks!
